### DISCUSSION TITLE
Errore "Conto di pagamento non trovato" su /c/COMPANY/issued_documents

### DISCUSSION THREAD

**Discussion by mauriziofonte** [fattureincloud/api#121] [Q&A] [answered]
Buongiorno a tutti,

in fase di creazione di una fattura stiamo ricevendo questo errore:

```
Client error: POST https://api-v2.fattureincloud.it/c/COMPANY/issued_documents resulted in a 422 Unprocessable Entity response:
{"error":{"message":"Conto di pagamento non trovato.","validation_result":null}}
```

Riporto un riassunto di codice utilizzato per creare il documento:

```php
<?php

// create the Items List Array
$ficProducts = array_map(function (array $invoiceLine) {
	$ficProduct = new \FattureInCloud\Model\IssuedDocumentItemsListItem([
		'qty' => 1,
		'code' => $invoiceLine['transaction']->code,
		'name' => $invoiceLine['transaction']->name,
		'description' => $invoiceLine['transaction']->description,
		'net_price' => $invoiceLine['transaction']->net,
		'gross_price' => $invoiceLine['transaction']->gross,
		'vat' => new \FattureInCloud\Model\VatType([
			'id' => 0,
			'value' =>  22,
			'description' => 'IVA 22%',
		]),
		'discount' => 0,
		'category' => 'Categoria',
	]);

	return $ficProduct;
}, $invoiceLines);

// create the Payments List Array
$ficPayments = array_map(function ($invoiceLine) {
	$ficPayment = new \FattureInCloud\Model\IssuedDocumentPaymentsListItem([
		'due_date' => $invoiceLine['transaction']->created_at->toDateTime(),
		'paid_date' => $invoiceLine['transaction']->created_at->toDateTime(),
		'status' => \FattureInCloud\Model\IssuedDocumentStatus::PAID,
		'amount' => $invoiceLine['gross_paid_price'],
		'payment_account' => new \FattureInCloud\Model\PaymentAccount([
			'type' => \FattureInCloud\Model\PaymentAccountType::STANDARD,
			'virtual' => false
		]), // secondo me il problema è qui
	]);

	return $ficPayment;
}, $invoiceLines);

// create the Payment Method Details Array
$ficPaymentMethodDetails = array_map(function ($invoiceLine) {
	$ficPaymentMethodDetail = new \FattureInCloud\Model\PaymentMethodDetails([
		'title' => "Pagamento Stripe",
		'description' => "Transazione Stripe #{$invoiceLine['transaction']->stripe_id} del {$invoiceLine['transaction']->created_at->format('d/m/Y H:i')}",
	]);

	return $ficPaymentMethodDetail;
}, $invoiceLines);

// create the Payment Method instance
$paymentMethod = new \FattureInCloud\Model\PaymentMethod([
	'name' => 'Stripe',
	'type' => \FattureInCloud\Model\PaymentMethodType::STANDARD,
	'is_default' => true,
	'details' => $ficPaymentMethodDetails,
]);

// create the Issued Document instance with Items List, Payments List, and Payment Method
$ficDocument = new \FattureInCloud\Model\IssuedDocument([
	'entity' => $this->getFicEntity($user),
	'type' => \FattureInCloud\Model\IssuedDocumentType::INVOICE,
	'date' => Carbon::now()->toDateTime(),
	'year' => Carbon::now()->format('Y'),
	'currency' => new \FattureInCloud\Model\Currency([
		'id' => 'EUR',
		'symbol' => '€',
		'exchange_rate' => '1.00000',
	]),
	'language' => new \FattureInCloud\Model\Language([
		'code' => 'it',
		'name' => 'Italiano',
	]),
	'subject' => "Identificativo fattura",
	'visible_subject' => "Identificativo fattura",
	'items_list' => $ficProducts,
	'payments_list' => $ficPayments,
	'payment_method' => $paymentMethod,
	'e_invoice' => true,
	'rc_center' => 'Incasso Stripe',
	'show_payment_method' => true,
	'show_totals' => \FattureInCloud\Model\ShowTotalsMode::ALL,
]);

// create the Document
$ficCreateDocument = new \FattureInCloud\Model\CreateIssuedDocumentRequest();
$ficCreateDocument->setData($ficDocument);

try {
	$apiInstance = new \FattureInCloud\Api\IssuedDocumentsApi($this->guzzleClient, $this->ficConfig);
	$ficIssuedDocument = $apiInstance->createIssuedDocument($this->ficCompanyId, $ficCreateDocument);
	$test = $this->toArray($ficIssuedDocument);
	var_dump($test);
	exit();
} catch (\Exception $e) {
	// exception handler
}
```

Il problema è che nella documentazione non sono segnalati i possibili messaggi di errore, nè come andare a risolverli.

Penso di aver identificato il punto specifico del payload che manda in errore tutta la richiesta (commento nel codice: **// secondo me il problema è qui**).

Come si risolve questo, nello specifico? Abbiamo configurato un sotto-account di test parallelo a quello in essere, che usa l'API V1 per il sito in produzione.
Stiamo migrando alla API V2, ma non essendo presente un'area Staging per le API, evidentemente mancano delle configurazioni in questo account di test.

Nota importante: i pagamenti sono collezionati attraverso Stripe, non vanno a finire in un conto bancario. Nella fattura deve essere chiaro che il pagamento è stato finalizzato attraverso Stripe.

**Comment by marco-sacchi** [comment, marked as answer]
Come visibile nello step 6 di [questa guida](https://developers.fattureincloud.it/docs/guides/invoice-creation#6%EF%B8%8F%E2%83%A3-step-six-payments-list) è obbligatorio specificare l'`id` dell'account di pagamento quando lo stato è impostato su `PAID`.

Diversamente dalla precedente versione, sembra che i conti di saldo non possano essere creati contestualmente alla creazione di un documento, ma è presente [questa request](https://developers.fattureincloud.it/api-reference/#post-/c/-company_id-/settings/payment_accounts) per crearne uno e acquisire l'`id` da inserire poi nel documento.

**Reply by mauriziofonte** [reply]
Gent.mo Marco,

grazie per aver risposto esaustivamente alla domanda.

Purtroppo però, nasce un altro errore. Riporto qui il codice usato per interfacciarsi direttamente alla **POST** `/c/{company_id}/settings/payment_accounts`, che risponde 422 con **"The i dati field is required."**:

**Exception on getFicPaymentAccountEntity: Fatture In Cloud Invoicer replied with a 422 status code. RAW="{"error":{"message":"Invalid request.","validation_result":{"data":["The i dati field is required."]}}}"**

```php
try {
	$paymentAccountResponse = $this->remoteCall("/c/COMPANY/settings/payment_accounts/" . $this->ficPaymentAccountId);
	if (empty($paymentAccountResponse)) {
		// we need to create a new Payment Account
		$paymentAccountResponse = $this->remoteCall("/c/COMPANY/settings/payment_accounts", [
			"type" => \FattureInCloud\Model\PaymentAccountType::STANDARD,
		]);

		if (!is_array($paymentAccountResponse)) {
			throw new \Exception("Cannot create the Payment Account.");
		}
	}

	$paymentAccount = new \FattureInCloud\Model\PaymentAccount($paymentAccountResponse);
	return $paymentAccount;
} catch (\Exception $e) {
	$this->exception("Exception on getFicPaymentAccountEntity: {$e->getMessage()}", 0, $e);
}
```

Il punto è, vorremmo poter creare un account di pagamento che non ha dati IBAN, non sono necessari. Le fatture vengono emesse **dopo che i fondi sono già stati pagati e acquisiti su Stripe**. E' possibile?

P.S. L'errore è lo stesso anche se si manda questo payload come da esempio su [https://developers.fattureincloud.it/api-reference/#post-/c/-company_id-/settings/payment_accounts](https://developers.fattureincloud.it/api-reference/#post-/c/-company_id-/settings/payment_accounts):

```
[
	"id" => $this->ficPaymentAccountId,
	"type" => \FattureInCloud\Model\PaymentAccountType::BANK,
	"name" => "Indesa",
	"iban" => "IT17Q0051343200000003497636",
	"sia" => "T1234",
	"virtual" => false
]
```

**Reply by emajo** [reply]
Ciao @mauriziofonte l'errore che ricevi è dovuto a una struttura sbagliata del payload che stai cercando di mandare, corretto sarebbe:

```
[
    "data" => 
	[
		"type" => \FattureInCloud\Model\PaymentAccountType::BANK,
		"name" => "Indesa",
		"iban" => "IT17Q0051343200000003497636",
		"sia" => "T1234",
		"virtual" => false
	]
]
```
per evitare questi problemi ti consiglio di utilizzare i metodi della nostra SDK, in questo caso il metodo [Create Payment Account](https://github.com/fattureincloud/fattureincloud-php-sdk/blob/master/docs/Api/SettingsApi.md#createpaymentaccount).

Per quanto riguarda la creazione di un payment account senza iban è possibile farlo senza alcun problema.
